Feature/#99

t-nagatsuka-gr opened the pull request.

t-nagatsuka-gr committed:
#99 プレイヤークラスの作成

t-nagatsuka-gr committed:
#99 プレイヤークラス
・コメントのインデント
・行末コメントの削除

t-nagatsuka-gr committed:
#99 プレイヤークラスの作成
・メソッド名などの見直し
・インデント修正
・行末コメントの削除

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,52 @@
+    public enum DetermineValue
+    {
+        One = 1,
+        Two = 2
```
「役を判定するための値」が何なのかわかりません。
役を判定するために「何に使う」値なのか、を意識してコメントを記載してください。

また、1と一致した場合は「どういう結果になる」のか、2の場合はどうなのか、という点も考えて、変数名やコメントを決めてください。

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,52 @@
+    Dice dice = new Dice();
+
+    // サイコロの乱数を保持するための配列
+    private int[] diceValue = new int[3];
```
ダイスクラスを使い回して値だけint型で3つ保持するのではなく、ダイスクラスのインスタンスを3つ持つようにしてください。

a-kiyo-g-relations replied:
配列の場合は名前を複数形にするなど、配列だとわかりやすい命名を心がけてください。

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,52 @@
+    // 指定の数ループし、取得した値を保持
+    public void LoopCountNumber()
+    {
+        int count = 3;
```
全体通して言える事ですが、リテラルは基本的にNGです。
意味のある値は基本的に全て定数定義してください。
(カウンタの0など、一部のケースのみが例外です)

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    // サイコロの出目の役を判定する
+    public int DetermineDicPip(int[] numbers)
```
役を判定して返している値が何を指すのかわかりません。
定義してください。

a-kiyo-g-relations replied:
サイコロの出目は保持しているので、引数は不要です。

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,52 @@
+    // サイコロの出目の役を判定する
+    public int DetermineDicPip(int[] numbers)
+    {
+        if (numbers.Distinct().Count() == (int)DetermineValue.One)
```
全体的に「どういう判定をして、どういう条件の場合、どういう結果になるのか」がわかりづらいです。
詳細設計を参考にして、コメントを充実させてください。

t-nagatsuka-gr committed:
#99 プレイヤークラスの作成
・Constantsクラスの作成
 定数の追加
・Diceクラスの修正
 ランダムメソッド実行時0が取得されてしまうため修正
・Playerクラスの修正、メソッド追加

a-kiyo-g-relations commented on Constants.cs:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    // ゾロ目を値「1」として取得する
+    public JudgeValue GetZorome()
```
これらのラッパーメソッドは必要ですか？
直接Constants.JudgeValue.Zoromeなどの参照で良いと思います。

a-kiyo-g-relations commented on Dice.cs:
```diff
@@ -42,7 +42,7 @@ public void RollDice()
         Array diceValues = Enum.GetValues(typeof(Pip));
-        int randomIndex = random.Next(diceValues.Length);
+        int randomIndex = random.Next(1, diceValues.Length + 1);
         return randomIndex;
```
ここは本来diceValues[randomIndex]を返すべきだったと思います。
```c#
// int型の配列形式でenum型の定義値一覧を取得する
int[] diceValues = (int[])Enum.GetValues(typeof(Pip));
// 0以上diceValues.length未満の整数を取得する
int randomIndex = random.Next(diceValues.Length);
// 配列の要素をランダムに返す
return diceValues[randomIndex];
``` 
こうする事でenumが連番でなくても正しく動作します。

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,89 @@
+public class Player
+{
+    // サイコロの乱数を保持するための配列
+    private int[] diceValues = new int[3];
```
不要と指摘済みです。

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,89 @@
+    private int[] diceValues = new int[3];
+
+    // サイコロを振った結果の出目数値
+    private int value;
```
これも不要です。二重管理になるものは原則消してください。

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,89 @@
+
+    Dice diceOne = new Dice();
+    Dice diceTwo = new Dice();
+    Dice diceThree = new Dice();
```
変数3つではなく、配列として保持してください。

a-kiyo-g-relations reviewed (changes requested):
前回コメント済みの内容が直っていません。
再度確認してください。

a-kiyo-g-relations commented on Constants.cs:
```diff
@@ -0,0 +1,31 @@
+public class Constants
+{
+    // 出目の種類によって役を判断するための値 
+    public enum JudgeValue
```
~~指摘漏れていましたが、定数はstaticで定義してください。~~

~~staticな要素はインスタンスを作成しなくてもアクセスする事が可能です。~~
~~定数クラスのインスタンスを複数作成する必要はありませんので、原則として定数はstaticで定義します。~~

enumの場合staticは不要です。
自動的に静的要素として扱われるため、インスタンスを作成しなくてもConstants.JudgeValue.Zoromeのような形で参照可能です。
readonlyを使用する場合は上記のようにstaticを付けてください。

a-kiyo-g-relations replied:
設計書見直して気付きましたが、これだと後から実装する役の強さ判定に使えない気がします。
「役を判定するための出目の種類の定義」と「役そのものの定義」が必要になりそうです。

以前の指摘が少しズレていたので申し訳ないのですが、以下のような感じになるかと思います。
```c#
// 手役の種類と強さ順。Constantsに定義する。仮に「123」や「456」などの特殊役を実装したらここが増える
public enum 手役
{
    役なし,
    役あり,
    ゾロ目
}

// 出目の種類で役判定に使う数字。役判定ロジック内部でしか使わないので、Playerクラス内で定義すれば良い
private const int 役判定_ゾロ目 = 1
private const int 役判定_役あり = 2
// 役なしは使わないので不要
``` 
役判定ロジックの方にも関連コメント記載しておきます。

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,89 @@
+    // enum型をobject型を経由してint型に変換する
+    private int EnumToInt<I>(I value) where I : Enum
+    {
+        return (int)(object)value;
```
「型変換メソッド」というメモから作ったのかと思いますが、これは不要です。
必要なのはDiceクラスの配列をint型の配列に変換して返す処理です。

こんな形で、Diceクラスの配列をメンバとして持っているとします。
```c#
Dice[] dices

dice[0].GetValue() // 値(出目)は2
dice[1].GetValue() // 値(出目)は3
dice[2].GetValue() // 値(出目)は5
``` 
欲しいのは、dicesのそれぞれの出目をint型の配列として返すメソッドです。
```c#
public int[] GetDiceValues()
{
    dicesの出目の値だけを取り出し、int型の配列に変換する処理
}

GetDiceValues() // [2, 3, 5]という配列が返ってくる
``` 
これを実装しておけば、diceValuesを変数で保持しなくても処理を行う事ができます。
```c#
GetDiceValues().Distinct().Count() // 今のdiceValues.Distinct().Count()と同じ動きになる
```

a-kiyo-g-relations replied:
配列の型や値を変換する処理ですが、C#の場合は以下のように行うのが良さそうです。
```c#
配列.Select(v => 各要素に行う処理).ToArray();
``` 
例えば、以下は数値型の配列を100倍して、文字列型に変換する処理です。
```c#
int[] values = new int[] {1, 2, 3};
string[] results = values.Select(value => (value * 100).ToString()).ToArray();
results[0]; // 文字列の"100"
results[1]; // 文字列の"200"
results[2]; // 文字列の"300"
```

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,89 @@
+        if (compValue == EnumToInt(Constants.JudgeValue.Zorome))
+        {
+            int Zorome = EnumToInt(Constants.JudgeValue.Zorome);
+            return Zorome; // 出目が1種類 =「役：ゾロ目」
```
「役を判定するための出目の種類の定義」と「役そのものの定義」が必要になる、という件と合わせて確認してください。

このメソッドの戻り値はint型ではなく「役そのものの定義」のenum型の方が良いです。
理由として、ここで返す値は「単なる数値型」ではなく、手役が何かを指す「意味のあるデータ型」だからです。

イメージとしては以下のような感じです。
```c#
public 手役 ComparisonPipTypeCount()
{
    // 中略
    if (compValue == 役判定_ゾロ目)
    {
        return 手役.ゾロ目
    }
    else if (compValue == 役判定_役あり)
    {
        return 手役.役あり
    }
    else
    {
        return 手役.役なし
    }
}
```

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    // 重複した値を除去し、残った数値を取得する
+    public void GetValueDice()
```
「取得する」なので、戻り値はvoid型ではなくint型になるかと思います。

a-kiyo-g-relations replied:
このロジックはおそらくゾロ目の時に落ちます。

もっとシンプルに考えて良いです。
```c#
if (ゾロ目の場合)
{
    return サイコロの先頭の出目;
}
else if (役ありの場合)
{
    return 1つしかない出目;
}
else
{
    return 0;
}
``` 
役ありの場合、1つしかない出目をどう取るかが悩むところかもしれませんが、無理にLINQのwhereなど使わなくても、以下のような単純なループ処理でいいです。
```c#
foreach (int val in diceValues) // valに先頭から順に出目が入っていく
{
    diceValuesにvalと同じ値が何個あるかを取得する
    if (1個しかない)
    {
        return val;
    }
}
// ここには来ないはずだが、エラーが出るはずなので一応書いておく
return 0;
``` 
この方がシンプルですし、パッと見で何をしているかわかりやすいはずです。

a-kiyo-g-relations replied:
メソッド名(あとコメントも)は「ダイスの出目」ではなく「役の強さ(出目)」が適切ですね。

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    // 出目の種類の数によって役を判定する
+    public int ComparisonPipTypeCount()
```
メソッド名は「役を取得する」みたいなものが良いかと思います。

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,89 @@
+    Dice diceOne = new Dice();
+    Dice diceTwo = new Dice();
+    Dice diceThree = new Dice();
+    Constants constants = new Constants();
```
~~別記の通り、staticで定義すればnewは不要になります。~~
enumはstaticなので、newは不要です。

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,89 @@
+    Constants constants = new Constants();
+
+    // サイコロを振り、取得した出目の値を配列に保持
+    public void RollDiceHoldingValue()
```
メソッド名はRollDiceだけで良いかと思います。
「配列に保持」という部分は内部処理の話なので、メソッド名に付けなくても良いです。

サイコロが複数なので、RollDice**s**だとより良いですね。

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,89 @@
+    private int value;
+
+
+
```
無駄な改行は消しておいてください。

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,89 @@
+    {
+        return (int)(object)value;
+    }
+}
```
ファイル末には空行を入れてください。

t-nagatsuka-gr committed:
#99 プレイヤークラスの作成

t-nagatsuka-gr committed:
#99 プレイヤークラスの作成
・命名とコメントの修正
・IsDiceValue()の修正

t-nagatsuka-gr committed:
#99 プレイヤークラスの作成
・ファイル末に空行の追加

a-kiyo-g-relations commented on Constants.cs:
```diff
@@ -0,0 +1,13 @@
+    {
+        YakuNashi, // 出目が0種類 =「役：役なし」
+        YakuAri,   // 出目が2種類 =「役：役あり」
+        Zorome,    // 出目が1種類 =「役：ゾロ目」
```
こちらは出目の種類に関する情報は無くなって、純粋に「役なし/役あり/ゾロ目」を意味する定数になったかと思います。

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,72 @@
+﻿using System;
+using System.Linq;
+
+
```
特に意味のない複数の改行があります。
1つで良いです。

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,72 @@
+
+    private Dice[] dices = new Dice[] { new Dice(), new Dice(), new Dice() };
+
+    // 三つのサイコロを振る
```
細かいですが「3つの」という部分は不要です。
今後サイコロ5つに変わった場合に直す箇所が無意味に増えてしまいます。

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,72 @@
+        }
+    }
+
+    // 三つのサイコロをint型配列に格納する
```
「格納する」は誤りです。
「サイコロの出目をint型配列に変換する」などが正確です。

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,72 @@
+    public int[] DicesToInt()
+    {
+        int[] diceValues = dices.Select(dice => dice.GetValue()).ToArray();
+        return diceValues;
```
ここで一度代入する意味はあまりないので、そのままreturnで良いと思います。

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    // 取得した値と出目の種類と比較、重複しない値を取得する
+    public int IsDiceValue(int[] values)
```
引数渡しではなく、DicesToIntで取得した値を使えば良いかと思います。

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,72 @@
+    // 取得した値と出目の種類と比較、重複しない値を取得する
+    public int IsDiceValue(int[] values)
+    {
+        if (values.Distinct().Count() == values.Length)
```
それぞれの判定に説明を記載しておいてください。

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    // 取得した出目と出目の種類を比較し役を取得する
+    public Constants.ToDecideMeans IsHand(int[] number)
```
こちらも上記同様、引数渡しではなく、DicesToIntで取得した値を使えば良いかと思います。

a-kiyo-g-relations replied:
Isは一般的にboolean型を返すメソッドの名前です。

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,72 @@
+        {
+            return 0;
+        }
+        else if (values.Distinct().Count() == RoleZorome)
```
ここの判定ですが、IsHandの中でやっているゾロ目の判定処理と同じ記載になっています。
こうしていると、仮にゾロ目の判定方法が変わった場合、修正箇所が2箇所になってしまいます。

そのため、以下のように判定した方がより良いです。
```c#
if (IsHand() == Constants.ToDecideMeans.Zorome)
```

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,72 @@
+        {
+            return values.First();
+        }
+        foreach (int val in values)
```
ここはループ外に「役ありの場合」というif文を入れておいた方が良いです。
他に役が増えた場合でもループに入ってしまうため、処理に無駄があります。

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,72 @@
+    }
+}
+
+
```
行末の空行は1つで良いです。

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    // 三つのサイコロをint型配列に格納する
+    public int[] DicesToInt()
```
Intは単数系なので、複数形の命名の方がより良いです。
diceのvalueを取得しているので、DiceValuesなどで良いんじゃないかと思います。

a-kiyo-g-relations commented on Player.cs:
```diff
@@ -0,0 +1,72 @@
+        foreach (int val in values)
+        {
+            int count = values.Count(x => x == val);
+            if (count == RoleYakuAri)
```
ここのcountは「役ありか否か」ではなく「複数存在するか」なので、定数を使うのは間違いかと思います。
`count > 1`などが正しいはずです。

t-nagatsuka-gr committed:
#99 プレイヤークラスの作成
